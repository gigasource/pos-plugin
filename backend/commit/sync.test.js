const mockProducts = require('../../components2/OrderView/__test__/mock_products').mockProducts
const _ = require('lodash')
const ObjectId = require('bson').ObjectId

const { syncFactory } = require('../../test-utils')

const { stringify } = require('schemahandler/utils')

jest.setTimeout(60000)

//<editor-fold desc="prepare test">
const {
  prepareActionCommitTest
} = require('./actionCommit.prepare.test')
const {
  prepareOrderTest,
  checkOrderCreated
} = require('../order/order.prepare.test')
const {
  prepareOrderCommit
} = require('./orderCommit.prepare.test')
//</editor-fold>

let cmsList
let cmsMaster
let cmsClient

beforeAll(async () => {
  cmsList = await syncFactory('syncTest', 2)

  for (let _cms of cmsList) {
    prepareActionCommitTest(_cms)
    prepareOrderCommit(_cms)
    prepareOrderTest(_cms)
    _cms.triggerFeConnect()
  }

  cmsMaster = cmsList[0]
  cmsClient = cmsList[1]
})

beforeEach(async () => {
  for (let _cms of cmsList) {
    await _cms.orm('Order').remove({}).direct()
    await _cms.orm('Commit').remove({}).direct()
    await _cms.orm('Action').remove({}).direct()
  }
})

describe('Sync test', function () {
  it('Case 1: Create order + add product', async (done) => {
    cmsClient.feSocket.on('update-table', async function (order) {
      const orderInDbMaster = await cmsMaster.orm('Order').find({})
      const orderInDbClient = await cmsClient.orm('Order').find({})
      expect(stringify(orderInDbClient)).toMatchSnapshot()
      expect(orderInDbMaster).toEqual(orderInDbClient)
      done()
    })
    const doc = await cmsMaster.orm('Order').create({
      _id: new ObjectId(),
      table: 10,
      status: 'inProgress'
    })
    // add first item
    await cmsMaster.orm('Order').findOneAndUpdate({
      _id: doc._id
    }, {
      $push: {
        items: mockProducts[0]
      }
    }).commit('updateActiveOrder', { table: 10 })
    // add second item
    await cmsMaster.orm('Order').findOneAndUpdate({
      _id: doc._id
    }, {
      $push: {
        items: mockProducts[1]
      }
    }).commit('updateActiveOrder', { table: 10 })
  })

  it('Case 1a: Create order + add product from client', async (done) => {
    cmsMaster.feSocket.on('update-table', async function (order) {
      const orderInDbMaster = await cmsMaster.orm('Order').find({})
      const orderInDbClient = await cmsClient.orm('Order').find({})
      expect(stringify(orderInDbClient)).toMatchSnapshot()
      expect(orderInDbMaster).toEqual(orderInDbClient)
      done()
    })
    const doc = await cmsClient.orm('Order').create({
      _id: new ObjectId(),
      table: 10,
      status: 'inProgress'
    })
    // add first item
    await cmsClient.orm('Order').findOneAndUpdate({
      _id: doc._id
    }, {
      $push: {
        items: mockProducts[0]
      }
    }).commit('updateActiveOrder', { table: 10 })
    // add second item
    await cmsClient.orm('Order').findOneAndUpdate({
      _id: doc._id
    }, {
      $push: {
        items: mockProducts[1]
      }
    }).commit('updateActiveOrder', { table: 10 })
  })

  it('Case 2: Create order + add order + set attribute', async (done) => {
    cmsClient.feSocket.on('update-table', async function (order) {
      const orderInDbMaster = await cmsMaster.orm('Order').find({})
      const orderInDbClient = await cmsClient.orm('Order').find({})
      expect(stringify(orderInDbClient)).toMatchSnapshot()
      expect(orderInDbMaster).toEqual(orderInDbClient)
      done()
    })
    const doc = await cmsMaster.orm('Order').create({
      _id: new ObjectId(),
      table: 10,
      status: 'inProgress'
    })
    // add first item
    await cmsMaster.orm('Order').findOneAndUpdate({
      _id: doc._id
    }, {
      $push: {
        items: mockProducts[0]
      }
    }).commit('updateActiveOrder', { table: 10 })
    // set attribute for order
    await cmsMaster.orm('Order').findOneAndUpdate({
      _id: doc._id
    }, {
      $set: {
        vSum: 100
      }
    })
    // add second item
    await cmsMaster.orm('Order').findOneAndUpdate({
      _id: doc._id
    }, {
      $push: {
        items: mockProducts[1]
      }
    }).commit('updateActiveOrder', { table: 10 })
  })
})
