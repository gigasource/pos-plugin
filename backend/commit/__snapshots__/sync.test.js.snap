// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync test Case 1: Create order + add product 1`] = `
Array [
  Object {
    "_id": "ObjectID",
    "cancellationItems": Array [],
    "discounts": Array [],
    "items": Array [
      Object {
        "_id": "ObjectID",
        "category": "ObjectID",
        "id": 1,
        "layouts": Array [
          Object {
            "color": "#FBE4EC",
            "device": "pos",
            "order": 0,
          },
        ],
        "name": "Cola",
        "option": Object {
          "favorite": true,
        },
        "price": 98,
        "tax": 19,
      },
      Object {
        "_id": "ObjectID",
        "category": "ObjectID",
        "id": 2,
        "layouts": Array [
          Object {
            "color": "#FBE4EC",
            "device": "pos",
            "order": 1,
          },
        ],
        "name": "Pepsi",
        "option": Object {
          "favorite": true,
        },
        "price": 35,
        "tax": 0,
      },
    ],
    "payment": Array [],
    "printer": Array [],
    "status": "inProgress",
    "table": 10,
    "user": Array [],
    "vTaxGroups": Array [],
  },
]
`;

exports[`Sync test Case 1: Create order + add product 2`] = `Array []`;

exports[`Sync test Case 1a: Add more than 1 products 1`] = `
Array [
  Object {
    "_id": "ObjectID",
    "cancellationItems": Array [],
    "cashback": -133,
    "discounts": Array [],
    "items": Array [
      Object {
        "_id": "ObjectID",
        "course": 1,
        "groupPrinter": "",
        "id": 2,
        "name": "Pepsi",
        "price": 35,
        "printed": true,
        "quantity": 1,
        "sent": true,
        "tax": 0,
        "taxes": Array [
          0,
          null,
        ],
        "vSum": 35,
        "vTakeAway": false,
        "vTaxSum": Object {
          "0": Object {
            "gross": 35,
            "net": 35,
            "tax": 0,
          },
        },
      },
      Object {
        "_id": "ObjectID",
        "course": 1,
        "groupPrinter": "",
        "id": 1,
        "name": "Cola",
        "price": 98,
        "printed": true,
        "quantity": 1,
        "sent": true,
        "tax": 19,
        "taxes": Array [
          19,
          null,
        ],
        "vSum": 98,
        "vTakeAway": false,
        "vTaxSum": Object {
          "19": Object {
            "gross": 98,
            "net": 82.35,
            "tax": 15.65,
          },
        },
      },
      Object {
        "_id": "ObjectID",
        "course": 1,
        "groupPrinter": "",
        "id": 2,
        "name": "Pepsi",
        "price": 35,
        "printed": true,
        "quantity": 1,
        "sent": true,
        "tax": 0,
        "taxes": Array [
          0,
          null,
        ],
        "vSum": 35,
        "vTakeAway": false,
        "vTaxSum": Object {
          "0": Object {
            "gross": 35,
            "net": 35,
            "tax": 0,
          },
        },
      },
    ],
    "payment": Array [],
    "printer": Array [],
    "status": "inProgress",
    "table": "10",
    "takeAway": false,
    "tip": 0,
    "user": Array [],
    "vDate": "2021-01-21T17:00:00.000Z",
    "vSum": 133,
    "vTaxGroups": Array [],
    "vTaxSum": Object {
      "0": Object {
        "gross": 35,
        "net": 35,
        "tax": 0,
      },
      "19": Object {
        "gross": 98,
        "net": 82.35,
        "tax": 15.65,
      },
    },
  },
]
`;

exports[`Sync test Case 1a: Create order + add product from client 1`] = `
Array [
  Object {
    "_id": "ObjectID",
    "cancellationItems": Array [],
    "discounts": Array [],
    "items": Array [
      Object {
        "_id": "ObjectID",
        "category": "ObjectID",
        "id": 1,
        "layouts": Array [
          Object {
            "color": "#FBE4EC",
            "device": "pos",
            "order": 0,
          },
        ],
        "name": "Cola",
        "option": Object {
          "favorite": true,
        },
        "price": 98,
        "tax": 19,
      },
      Object {
        "_id": "ObjectID",
        "category": "ObjectID",
        "id": 2,
        "layouts": Array [
          Object {
            "color": "#FBE4EC",
            "device": "pos",
            "order": 1,
          },
        ],
        "name": "Pepsi",
        "option": Object {
          "favorite": true,
        },
        "price": 35,
        "tax": 0,
      },
    ],
    "payment": Array [],
    "printer": Array [],
    "status": "inProgress",
    "table": 10,
    "user": Array [],
    "vTaxGroups": Array [],
  },
]
`;

exports[`Sync test Case 2: Create order + add order + set attribute 1`] = `
Array [
  Object {
    "_id": "ObjectID",
    "cancellationItems": Array [],
    "discounts": Array [],
    "items": Array [
      Object {
        "_id": "ObjectID",
        "category": "ObjectID",
        "id": 1,
        "layouts": Array [
          Object {
            "color": "#FBE4EC",
            "device": "pos",
            "order": 0,
          },
        ],
        "name": "Cola",
        "option": Object {
          "favorite": true,
        },
        "price": 98,
        "tax": 19,
      },
      Object {
        "_id": "ObjectID",
        "category": "ObjectID",
        "id": 2,
        "layouts": Array [
          Object {
            "color": "#FBE4EC",
            "device": "pos",
            "order": 1,
          },
        ],
        "name": "Pepsi",
        "option": Object {
          "favorite": true,
        },
        "price": 35,
        "tax": 0,
      },
    ],
    "payment": Array [],
    "printer": Array [],
    "status": "inProgress",
    "table": 10,
    "user": Array [],
    "vSum": 100,
    "vTaxGroups": Array [],
  },
]
`;
